### DISCUSSION TITLE
How to update Eglot post Emacs 29

### DISCUSSION THREAD

**Discussion by refaelsh** [joaotavora/eglot#1241] [Q&A] [answered]
Since Eglot is now part of Emacs 29, I have several questions please:
1. How to update the built-in Eglot in Emacs 29?
2. Will `(use-package eglot)` continue to work as usual?
3. What will happen to Github's repo of Eglot?

**Comment by joaotavora** [comment, marked as answer]
> How to update the built-in Eglot in Emacs 29?

The safest way is to type `M-x eglot-upgrade-eglot` into Emacs 29.  That command also works in Emacs versions older than 29, provided that  `eglot.el` is at least 1.16.  For Eglot 1.15, the command is named `eglot-update`.

> 2. Will (use-package eglot) continue to work as usual?

Depends on what part of that command you were relying on:  
* If you relied on that command to live in your configuration and be able to bootstrap an entirely new configuration (say, in a new machine you're setting up) and rev up Eglot to its latest version, than the answer is *NO, it will not continue to work as usual*. In Emacs 29, the command will silently complete but Eglot will stay at an ever-ageing 1.12.29 version, which is the built-in version that ships with Emacs 29.
* If you _never_ used that command to bootstrap new systems and expect Eglot to be upgraded to latest, then you may continue to use it. In this case, you may manually use `eglot-upgrade-eglot` to advance the Eglot version manually to the latest available one.  
 
The remaining parts of the `use-package` form, where you presumably configure certain parts of Eglot, will most likely continue to function.

> 3. What will happen to Github's repo of Eglot?

For the time being, nothing except what is already [described in the README.md](https://github.com/joaotavora/eglot#status-of-this-github-repository)

**Reply by refaelsh** [reply]
> The safest way is to type M-x eglot-upgrade-eglot into Emacs 29. That command also works in Emacs versions older than 29, provided that eglot.el is at least 1.16. For Eglot 1.15, the command is named eglot-update.

And how to do it (keep Eglot up-to-date) via init.el?

**Reply by joaotavora** [reply]
> And how to do it (keep Eglot up-to-date) via init.el?

You can also use `eglot-upgrade-eglot` non-interactively.  Just include this line `(eglot-upgrade-eglot)` in your `init.el` file.  This will refresh from the archives if not already done in the session and pull in and compile a new version of Eglot if it detects one.

**Reply by refaelsh** [reply]
> This will refresh from the archives

What do you mean by that? From where will it be updated? Not from this GitHub repo?

**Reply by refaelsh** [reply]
> You can also use eglot-upgrade-eglot non-interactively. Just include this line (eglot-upgrade-eglot) in your init.el file. This will refresh from the archives if not already done in the session and pull in and compile a new version of Eglot if it detects one.

Is there a way to do it with a package manager and not with a function specific to Eglot?

**Reply by refaelsh** [reply]
> You can also use eglot-upgrade-eglot non-interactively. Just include this line (eglot-upgrade-eglot) in your init.el file. This will refresh from the archives if not already done in the session and pull in and compile a new version of Eglot if it detects one.

Using the function you suggested, will make Emacs load slower because it will check for Eglot update every time. Is there some other solution? LIGHT BULB: is there a generic way in Emacs to update built-in packages?

**Reply by joaotavora** [reply]
> What do you mean by that? From where will it be updated? Not from this GitHub repo?

No, not from his GitHub repo.  From the GNU ELPA archives.  In fact, only non-standard package-managers ever fetched from this repo.  Have you read the README.me of this GitHub repo already?  It explains this repo is _not_ the upstream anymore.  
Eglot lives in Emacs.

> Is there a way to do it with a package manager and not with a function specific to Eglot?

Depends on your package manager.  I think yes, even with the built-in package manager of Emacs, there is a way to configure it or ask it to always update packages.  But I don't have a lot of information on that.  You should consult the documentation or ask in help-gnu-emacs@gnu.org.  I believe other package managers, like `straight.el` may offer the "update to latest" functionality by default.

> Using the function you suggested, will make Emacs load slower because it will check for Eglot update every time.

Look, either you want to check for an update on startup or you don't.  If you do, you need to wait the time to do the check.  If you don't, then you don't need to wait the time, but you also don't know if there's an updated version.  There are no instant telepathic checks for new version.

> Why was it decided to make Eglot part of Emacs? 

This is explained in the README.md.

> Would it not be better to leave it as a package, so it can be updated easily and not wait for an Emacs update?

Presumably, `M-x eglot-upgrade-eglot` is not hard.  It could be even simpler, and most of all backward compatible, but there was a big discussion and people decided that upgrading built-in packages is dangerous, even though that's what users like you and others are probably used to doing for Eglot.

**Reply by refaelsh** [reply]
> Look, either you want to check for an update on startup or you don't. If you do, you need to wait the time to do the check. If you don't, then you don't need to wait the time, but you also don't know if there's an updated version. There are no instant telepathic checks for new version.

Yes, I understand. Thank you :-)

**Reply by refaelsh** [reply]
> This is explained in the README.md.

You mean this: "From a development perspective, moving to core allows us to work on Eglot in tandem with other related packages already in Emacs, such as Flymake, ElDoc, Xref, Project."?

**Reply by refaelsh** [reply]
> but there was a big discussion

Can you please tell me where? I would really much like to read it.

**Reply by joaotavora** [reply]
> You mean this: "From a development perspective, moving to core allows us to work on Eglot in tandem with other related packages already in Emacs, such as Flymake, ElDoc, Xref, Project."?

Yes, exactly.

> Can you please tell me where? I would really much like to read it.

It's on the bug tracker, subject "Not easy at all to upgrade :core packages like Eglot".  It will be a very tiring and confusing read, and the only reason I can imagine someone wanting to read it now is for the popcorn. Even though I don't agree with the conclusion, I'm glad that it's done. So I will not say any more about it.

**Reply by refaelsh** [reply]
I understand now. Thank you, @joaotavora, for your replies.

**Reply by refaelsh** [reply]
> someone wanting to read it now is for the popcorn

I like popcorn!

**Reply by refaelsh** [reply]
> It's on the bug tracker, subject "Not easy at all to upgrade :core packages like Eglot"

I can't find it here: https://debbugs.gnu.org/cgi/pkgreport.cgi?package=emacs. I tried the search box with the subject you wrote and several other keywords. Can you please help?

**Reply by joaotavora** [reply]
> I like popcorn!

Yes, I figured.

> Can you please help?

No.

**Comment by refaelsh** [comment]
> Since Eglot is now part of Emacs 29, I have several questions please:
> 
>     1. How to update the built-in Eglot in Emacs 29?
> 
>     2. Will `(use-package eglot)` continue to work as usual?
> 
>     3. What will happen to Github's repo of Eglot?

4. Why was it decided to make Eglot part of Emacs? Would it not be better to leave it as a package, so it can be updated easily and not wait for an Emacs update?
